cookbooks/bcpc-hadoop/templates/default/hdp_hadoop-env.sh.erb

```diff
@@ -48,8 +48,16 @@ export JSVC_HOME=/usr/lib/bigtop-utils
 # On secure datanodes, user to run the datanode as after dropping privileges
 export HADOOP_SECURE_DN_USER=hdfs
 
+<% datanode_heap = [node["bcpc"]["hadoop"]["datanode"]["xmx"]["max_size"], (node["memory"]["total"].to_i * node["bcpc"]["hadoop"]["datanode"]["xmx"]["max_ratio"]/1024).floor].min %>
+<% datanode_newsize = [(0.125*datanode_heap).ceil, 3072].min %>
+<% namenode_heap = [node["bcpc"]["hadoop"]["namenode"]["xmx"]["max_size"], (node["memory"]["total"].to_i * node["bcpc"]["hadoop"]["namenode"]["xmx"]["max_ratio"]/1024).floor].min %>
+<% namenode_newsize = [(0.125*namenode_heap).ceil, 3072].min %>
+
+export HADOOP_DATANODE_OPTS="$HADOOP_DATANODE_OPTS -Xms<%= datanode_heap %>m -Xmx<%= datanode_heap %>m -XX:NewSize=<%= datanode_newsize %>m -XX:MaxNewSize=<%= datanode_newsize %>m -XX:PermSize=128m -XX:MaxPermSize=256m <%= node["bcpc"]["hadoop"]["datanode"]["gc_opts"] %>"
+export HADOOP_NAMENODE_OPTS="$HADOOP_NAMENODE_OPTS -Xms<%= namenode_heap %>m -Xmx<%= namenode_heap %>m -XX:NewSize=<%= namenode_newsize %>m -XX:MaxNewSize=<%= namenode_newsize %>m -XX:PermSize=128m -XX:MaxPermSize=256m <%= node["bcpc"]["hadoop"]["namenode"]["gc_opts"] %>"
```

Comment:
Good call, I was testing on hardware. I can scale perm size and max perm size from a floor of 32/64MB to the recommended sizes above.